src/CircularArrays.jl

```diff
@@ -26,19 +27,32 @@ Alias for [`CircularArray{T,1}`](@ref).
 """
 const CircularVector{T} = CircularArray{T, 1}
 
-@inline clamp_bounds(arr::CircularArray, I::Tuple{Vararg{Int}})::AbstractArray{Int, 1} = map(dim -> mod1(I[dim], size(arr.data, dim)), eachindex(I))
+# Copied from a method Base.mod, for compatibility with Julia version < 1.3,
+# where this method is not defined
+_mod(i::Integer, r::AbstractUnitRange{<:Integer}) = mod(i-first(r), length(r)) + first(r)
+@inline clamp_bounds(arr::CircularArray, I::Tuple{Vararg{Int}})::AbstractArray{Int, 1} = map(Base.splat(_mod), zip(I, axes(arr.data)))
 
 CircularArray(def::T, size) where T = CircularArray(fill(def, size))
 
-@inline Base.getindex(arr::CircularArray, i::Int) = @inbounds getindex(arr.data, mod1(i, size(arr.data, 1)))
-@inline Base.setindex!(arr::CircularArray, v, i::Int) = @inbounds setindex!(arr.data, v, mod1(i, size(arr.data, 1)))
+@inline Base.getindex(arr::CircularArray, i::Int) = @inbounds getindex(arr.data, mod(i, Base.axes1(arr.data)))
+@inline Base.setindex!(arr::CircularArray, v, i::Int) = @inbounds setindex!(arr.data, v, mod(i, Base.axes1(arr.data)))
 @inline Base.getindex(arr::CircularArray, I::Vararg{Int}) = @inbounds getindex(arr.data, clamp_bounds(arr, I)...)
 @inline Base.setindex!(arr::CircularArray, v, I::Vararg{Int}) = @inbounds setindex!(arr.data, v, clamp_bounds(arr, I)...)
 @inline Base.size(arr::CircularArray) = size(arr.data)
+@inline Base.axes(arr::CircularArray) = axes(arr.data)
+
+@inline Base.checkbounds(::CircularArray, _...) = nothing
+
+@inline _similar(arr::CircularArray, ::Type{T}, dims) where T = CircularArray(similar(arr.data,T,dims))
+@inline Base.similar(arr::CircularArray, ::Type{T}, dims::Tuple{Base.DimOrInd, Vararg{Base.DimOrInd}}) where T = _similar(arr,T,dims)
+# Ambiguity resolution with Base
+@inline Base.similar(arr::CircularArray, ::Type{T}, dims::Tuple{Int64,Vararg{Int64}}) where T = _similar(arr,T,dims)
+# Ambiguity resolution with OffsetArrays, for the case similar(arr) where arr.data::OffsetArray
+@inline Base.similar(arr::CircularArray, ::Type{T}, dims::Tuple{OffsetAxis, Vararg{OffsetAxis}}) where T = _similar(arr,T,dims)
```

Comment:
Is the hard dependency on OffsetArrays for this really necessary? I doubt that it would have to be. Why would this ambiguity resolution be needed at all? I'd like to avoid this hard dependency if possible, it's fine using the package for testing though.

Reply:
The reason ambiguity resolution is needed is the `similar` method mentioned here: JuliaArrays/OffsetArrays.jl#87.
Without changes to `Base` or `OffsetArrays`, the only solutions I can see are:
- a dependency on `OffsetArrays`
- copy-pasting the `OffsetAxis` definition from `OffsetArrays`
- defining our own axis type as suggested here https://github.com/JuliaArrays/OffsetArrays.jl/issues/87#issuecomment-560375282, and defining `similar` just for those. This would mean `axes` converts the underlying arrays `axes` to a custom range type, and `similar` does the opposite conversion to send the axes to the underlying array. This would solve the case `similar(a::CircularArray)` where `a.data` is offset, but not a direct call from user code like `similar(a::CircularArray, -2:2, -2:2)` (which would dispatch on the `OffsetArrays` definition and thus fail to produce a circular array). Perhaps I'm misunderstanding how this mechanism should be used. Maybe @timholy can elucidate?

Otherwise, I think changing `Base` and `OffsetArrays` as suggested here https://github.com/JuliaArrays/OffsetArrays.jl/issues/87#issuecomment-560376691 would solve this problem too.

Reply:
This sounds like an issue that needs to be fixed in OffsetArrays. I am not a fan of adding workarounds for problems in other packages in here.

Reply:
Would it be possible to have it more general [like in BlockArrays.jl](https://github.com/JuliaArrays/BlockArrays.jl/blob/45213476e35af2bce0f7dadde6209fcc5cfa2c33/src/abstractblockarray.jl#L35-L38), perhaps?

Reply:
If it's absolutely necessary, I'd rather copy the `OffsetAxis` definition over. It's just a tuple, after all.

Reply:
https://github.com/JuliaArrays/OffsetArrays.jl/pull/90 shows an example of defining your own axis type. You can specialize `axes` for the `axes`, as done in that PR.

Reply:
I may be missing something, but as far as I can see, this OffsetArrays example works because it defines `similar` not only for its custom index types, but also for "standard" types appearing in the `OffsetAxis` union, because `getindex(A,I...)` generally calls `similar` with `I`'s axes. We similarly can't avoid defining `similar` on the "standard" types, and then we're back with the ambiguity errors. If I only define `similar` for a custom axis type, then indexing `::CircularArray[::UnitRange]` fails to produce a `CircularArray` (because it calls a `similar` method from base).
At this point, I don't see how a custom index type is helpful at all. For now I will resolve the ambiguity as in `BlockArrays.jl`.